alf/utils/value_ops.py

```diff
@@ -271,3 +271,124 @@ def weighted_cumulative_td_fn(accumulated_td, weights_td_is_last):
         advantages = tf.transpose(a=advantages)
 
     return tf.stop_gradient(advantages)
+
+
+def vtrace_returns_and_advantages_impl(importance_ratio_clipped,
+                                       rewards,
+                                       values,
+                                       step_types,
+                                       discounts,
+                                       td_lambda=1,
+                                       time_major=True):
+    """Actual implementation after getting importance_ratios
+
+    Args:
+        importance_ratio_clipped (Tensor): shape is [T, B] vtrace IS weights.
+        rewards (Tensor): shape is [T, B] (or [T]) representing rewards.
+        values (Tensor): shape is [T,B] (or [T]) representing values.
+        step_types (Tensor): shape is [T,B] (or [T]) representing step types.
+        discounts (Tensor): shape is [T, B] (or [T]) representing discounts.
+        td_lambda (float): A scalar between [0, 1]. It's used for variance
+            reduction in temporal difference.
+        time_major (bool): Whether input tensors are time major.
+            False means input tensors have shape [B, T].
+
+    Returns:
+        Two tensors with shape [T-1, B] representing returns and advantages.
+        Shape is [B, T-1] when time_major is false.
+    """
+    if not time_major:
+        importance_ratio_clipped = tf.transpose(a=importance_ratio_clipped)
+        discounts = tf.transpose(a=discounts)
+        rewards = tf.transpose(a=rewards)
+        values = tf.transpose(a=values)
+        step_types = tf.transpose(a=step_types)
+
+    importance_ratio_clipped = importance_ratio_clipped[:-1]
+    rewards = rewards[1:]
+    next_values = values[1:]
+    final_value = values[-1]
+    values = values[:-1]
+    discounts = discounts[1:]
+    step_types = step_types[:-1]
+    is_lasts = tf.cast(tf.equal(step_types, StepType.LAST), tf.float32)
+
+    tds = (importance_ratio_clipped *
+           (rewards + discounts * next_values - values))
+    weighted_discounts = discounts * importance_ratio_clipped * td_lambda
+
+    def vs_target_minus_vs_fn(vs_target_minus_vs, params):
+        weighted_discount, td, is_last = params
+        return (1 - is_last) * (td + weighted_discount * vs_target_minus_vs)
+
+    vs_target_minus_vs = tf.scan(
+        fn=vs_target_minus_vs_fn,
+        elems=(weighted_discounts, tds, is_lasts),
+        initializer=tf.zeros_like(final_value),
+        reverse=True,
+        back_prop=False)
+
+    returns = (1 - is_lasts) * vs_target_minus_vs + values
```

Comment:
it seems that it's not necessary to multiply (1 - is_lasts) here, as it's already done at line 322